tests/Unit/Options/Test_Auto.cpp

```diff
@@ -0,0 +1,153 @@
+#include "Utilities/TMPL.hpp"
+
+namespace {
+template <typename T>
+void test_instance(Options::Auto<T> value,
+                   const std::optional<T> expected) noexcept {
+  CHECK(static_cast<const std::optional<T>&>(value) == expected);
+  CHECK(static_cast<std::optional<T>>(std::move(value)) == expected);
+}
+
+void test_class() noexcept {
+  test_instance(Options::Auto<int>{}, std::optional<int>{});
+  test_instance(Options::Auto<int>{3}, std::optional<int>{3});
+  test_instance(Options::Auto<std::unique_ptr<int>>{},
+                std::optional<std::unique_ptr<int>>{});
+  {
+    Options::Auto<std::unique_ptr<int>> value{std::make_unique<int>(3)};
+    const std::optional<std::unique_ptr<int>>& contained = value;
+    CHECK((contained and *contained and **contained == 3));
+    const std::optional<std::unique_ptr<int>> extracted = std::move(value);
+    CHECK((extracted and *extracted and **extracted == 3));
+  }
+
+  CHECK(get_output(Options::Auto<int>{}) == "Auto");
+  CHECK(get_output(Options::Auto<int>{3}) == "3");
+  CHECK(get_output(Options::Auto<std::vector<int>>{{1, 2}}) ==
+        get_output(std::vector<int>{1, 2}));
+}
+
+struct Derived;
+
+struct Base {
+  Base() = default;
+  Base(const Base&) = default;
+  Base& operator=(Base&&) = default;
+  Base(Base&&) = default;
+  Base& operator=(const Base&) = default;
+  virtual ~Base() = default;
+  using creatable_classes = tmpl::list<Derived>;
+};
+
+struct Derived : Base {
+  using options = tmpl::list<>;
+  static constexpr Options::String help = "halp";
+};
+
+template <typename T>
+void check_create(const std::string& creation_string,
+                  const std::optional<T>& expected) noexcept {
+  CAPTURE(creation_string);
+  CHECK(static_cast<std::optional<T>>(
+            TestHelpers::test_creation<Options::Auto<T>>(creation_string)) ==
+        expected);
+}
+
+void test_parsing() noexcept {
+  check_create<int>("3", 3);
+  check_create<int>("Auto", {});
+
+  check_create<std::string>("Auto", {});
+  check_create<std::string>("Not Auto", "Not Auto");
+
+  check_create<std::unique_ptr<Base>>("Auto", {});
+  {
+    const std::optional<std::unique_ptr<Base>> created =
+        TestHelpers::test_creation<Options::Auto<std::unique_ptr<Base>>>(
+            "Derived");
+    CHECK(created);
+    CHECK(*created);
+  }
+
+  check_create<std::vector<int>>("Auto", {});
+  check_create<std::vector<int>>("", std::vector<int>{});
+  check_create<std::vector<int>>("[1, 2, 3]", std::vector<int>{1, 2, 3});
+}
+
+/// [example_class]
+class ExampleClass {
+ public:
+  ExampleClass() = default;
+
+  struct AutoArg {
+    using type = Options::Auto<int>;
+#if defined(__GNUC__) && !defined(__clang__) && __GNUC__ >= 8 && __GNUC__ < 11
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif  // defined(__GNUC__) && !defined(__clang__) && __GNUC__ => 8 && __GNUC__ < 11
+    static type default_value() noexcept { return {}; }
```

Comment:
- Do you have an idea why GCC doesn't like this instantiation, but is fine with the one in the CCE code? It's a bit unfortunate that the pragmas get printed to the documentation snippet.
- You add pragmas for GCC 8--9 in Auto.hpp, and for GCC 8--10 here. Is that on purpose?

Please squash directly if you end up making any changes.

Reply:
I have no idea why different version of GCC complain about different variations on that line.

Good point about the documentation.  I've moved the pragmas outside the snippet.

Reply:
So GCC 10 indeed complains about the one line but not the other? that's weird